chemprop/uncertainty/calibrator.py

```diff
@@ -236,12 +299,85 @@ def apply(self, uncs: Tensor) -> Tensor:
+
+    .. math::
+        \Pr \left(
+            \hat{\mathcal C}_{\text{in}}(X) \subseteq \mathcal Y \subseteq \hat{\mathcal C}_{\text{out}}(X)
+        \right) \geq 1 - \alpha,
+
+    where the in-set :math:`\hat{\mathcal C}_\text{in}` is contained by the set of true labels :math:`\mathcal Y` and
+    :math:`\mathcal Y` is contained within the out-set :math:`\hat{\mathcal C}_\text{out}`.
+
+    Parameters
+    ----------
+    alpha: float
+        The error rate, :math:`\alpha \in [0, 1]`
+
+    References
+    ----------
+    .. [1] Cauchois, M.; Gupta, S.; Duchi, J.; "Knowing What You Know: Valid and Validated Confidence Sets
+        in Multiclass and Multilabel Prediction." arXiv Preprint 2020, https://arxiv.org/abs/2004.10181
+    """
+
+    def __init__(self, alpha: float):
+        super().__init__()
+        self.alpha = alpha
+        if not 0 <= self.alpha <= 1:
+            raise ValueError(f"arg `alpha` must be between 0 and 1. got: {alpha}.")
+
+    @staticmethod
+    def nonconformity_scores(preds: Tensor):
+        r"""
+        Compute nonconformity score as the negative of the predicted probability.
+
+        .. math::
+            s_i = -\hat{f}(X_i)_{Y_i}
+        """
+        return -preds
+
     def fit(self, uncs: Tensor, targets: Tensor, mask: Tensor) -> Self:
+        if targets.shape[1] < 2:
+            raise ValueError(
+                f"the number of tasks should be larger than 1! got: {targets.shape[1]}."
+            )
+
+        has_zeros = torch.any(targets == 0, dim=1)
+        index_zeros = targets[has_zeros] == 0
+        scores_in = self.nonconformity_scores(uncs[has_zeros])
+        masked_scores_in = scores_in * index_zeros.float() + torch.where(
+            index_zeros, torch.zeros_like(scores_in), torch.tensor(float("inf"))
+        )
+        calibration_scores_in = torch.min(
+            masked_scores_in.masked_fill(~mask, float("inf")), dim=1
+        ).values
+
+        has_ones = torch.any(targets == 1, dim=1)
+        index_ones = targets[has_ones] == 1
+        scores_out = self.nonconformity_scores(uncs[has_ones])
+        masked_scores_out = scores_out * index_ones.float() + torch.where(
+            index_ones, torch.zeros_like(scores_out), torch.tensor(float("-inf"))
+        )
+        calibration_scores_out = torch.max(
+            masked_scores_out.masked_fill(~mask, float("-inf")), dim=1
+        ).values
+
+        self.tout = torch.quantile(
+            calibration_scores_out, 1 - self.alpha / 2, interpolation="higher"
+        )
+        self.tin = torch.quantile(calibration_scores_in, self.alpha / 2, interpolation="higher")
         return self
 
     def apply(self, uncs: Tensor) -> Tensor:
-        return
+        scores = self.nonconformity_scores(uncs)
+
+        cal_preds_in = (scores <= self.tin).int()
+        cal_preds_out = (scores <= self.tout).int()
+        cal_preds_in_out = torch.cat((cal_preds_in, cal_preds_out), dim=1)
```

Comment:
Would it make more sense to stack the in and out in a new dimension? Then the 1st dimension length would stay the number of tasks instead of two times the number of tasks. The 2nd dimension (which we use for uncertainty) would be used for the in vs. out sets.